src/mods/stage_edits.cpp

```diff
@@ -0,0 +1,105 @@
+};
+
+enum class GoalType {
+    Any = 0,
+    Blue = 1,
+    Green = 2,
+    Red = 3,
+};
+
+static patch::Tramp<decltype(&mkb::load_stagedef)> load_stagedef_tramp;
+
+u32 count_goals(u32 type) {
+    u32 count = 0;
+    for (u32 i = 0; i < mkb::stagedef->goal_count; i++) {
+        if (mkb::stagedef->goal_list[i].type == type) count++;
+    }
+    return count;
+}
+
+void init() {
+    patch::hook_function(load_stagedef_tramp, mkb::load_stagedef, [](u32 stage_id) {
+        load_stagedef_tramp.dest(stage_id);
+        if (pref::get(pref::BoolPref::ReverseMode) && mkb::stagedef->goal_count > 0) {
+            u32 goal_num = mkb::rand() % mkb::stagedef->goal_count;
+            u8 type = pref::get(pref::U8Pref::ReverseGoalType);
+            if (type != 0) {
+                u32 count = count_goals(type - 1);
+                if (count > 0) {
+                    u32 target = mkb::rand() % count;
+                    for (u32 i = 0; i < mkb::stagedef->goal_count; i++) {
+                        if (mkb::stagedef->goal_list[i].type != type - 1) continue;
+                        if (target == 0) {
+                            goal_num = i;
+                        } else {
+                            target--;
+                        }
+                    }
+                }
+            }
+            // switch goal and start
+            float x = mkb::stagedef->start->position.x;
+            float y = mkb::stagedef->start->position.y;
+            float z = mkb::stagedef->start->position.z;
+            s16 sx = mkb::stagedef->start->rotation.x;
+            s16 sy = mkb::stagedef->start->rotation.y;
+            s16 sz = mkb::stagedef->start->rotation.z;
+
+            mkb::stagedef->start->position.x = mkb::stagedef->goal_list[goal_num].position.x;
+            mkb::stagedef->start->position.y = mkb::stagedef->goal_list[goal_num].position.y + 0.5;
+            mkb::stagedef->start->position.z = mkb::stagedef->goal_list[goal_num].position.z;
+            mkb::stagedef->start->rotation.x = mkb::stagedef->goal_list[goal_num].rotation.x;
+            mkb::stagedef->start->rotation.y =
+                mkb::stagedef->goal_list[goal_num].rotation.y - 32766;
+            mkb::stagedef->start->rotation.z = mkb::stagedef->goal_list[goal_num].rotation.z;
+
+            mkb::stagedef->goal_list[goal_num].position.x = x;
+            mkb::stagedef->goal_list[goal_num].position.y = y - 0.5;
+            mkb::stagedef->goal_list[goal_num].position.z = z;
+            mkb::stagedef->goal_list[goal_num].rotation.x = sx;
+            mkb::stagedef->goal_list[goal_num].rotation.y = sy - 32766;
+            mkb::stagedef->goal_list[goal_num].rotation.z = sz;
+        } else if (BananaType(pref::get(pref::U8Pref::BananaType)) == BananaType::Golden) {
+            // disable goals somehow
+            for (u32 i = 0; i < mkb::stagedef->goal_count; i++) {
+                mkb::stagedef->goal_list[i].position.y = -10000;
```

Comment:
I'm worried this might exacerbate the issue where the ball can jitter on stages with rotating objects. Can you make sure the issue isn't much worse on a stage like Coin Slots?

Reply:
I tried on challenge mode coin slots and the jittering was actually lessened, I was kind of disappointed because I was hoping I'd be flying all over the place. I'll test it some more before merging though

Reply:
Makes sense it wouldn't be worse tbh, translation isn't as bad as rotation because it wouldn't affect ball velocity